File: problems/data-structures-n-algorithms/medium/3.js
// 3. Longest Substring Without Repeating Characters
// https://leetcode.com/problems/longest-substring-without-repeating-characters/

/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    var sLen = s.length;
    var len=0;
    var max = 0;
    var map = {};
    for(var i = 0; i < sLen; i++){
        if(!map.hasOwnProperty(s[i])){
            map[s[i]] = i;
            len++;
        }else{
            if(len > max){
                max = len;
            }
            len = 0;
            i = map[s[i]];
            map = {};


        }
    }

    if(len > max){
        max = len;
    }

    return max;
};

File: problems/data-structures-n-algorithms/medium/3.java
// 3. Longest Substring Without Repeating Characters
// https://leetcode.com/problems/longest-substring-without-repeating-characters/

import java.util.Arrays;

class S3 {
	public static void main (String args[]) {
		// String s = "abcabcbb";
		// String s = "bbbbb";
		String s = "pwwkew";
		int res = lengthOfLongestSubstring(s);
		System.out.println(res);
	}

	public static int lengthOfLongestSubstring (String s) {
		int len = s.length();
		int[] map = new int[128];
		Arrays.fill(map, -1);

		int max = 0;
		int start = 0;
		for (int i = 0; i < len; i++) {
			if (map[s.charAt(i)] != -1) {
				while (s.charAt(start) != s.charAt(i)) {
					map[s.charAt(start)] = -1;
					start++;
				}
				start++;
			}
			map[s.charAt(i)] = i;
			if (i - start + 1 > max) {
				max = i - start + 1;
			}
		}
		return max;
	}
}